ve1ld/vyasa
# Fix mounting woes by rendering content as live_component within DM

basically took the whole day, a bunch of refactoring has happened 

# How to Verify / Test

1. navigate through the thing, if you change the mode then navigate into other parts of the content (e.g. :show_sources -> :show_chapters -> :show_verses) then the ControlPanel mode should be the same, no dismount and remounting or anything either

2. handshake gets done properly. The handshake will init when you're in the :show_verses content view and only then does the audio load. Note that once the audio is loaded, any content navigation will never kill the audio playback either. In fact the audio playback will continue and scribing and all that fun stuff will still happen


# Workplan Notes

```
***** Approach 2: ContentLiveview
:LOGBOOK:
CLOCK: [2024-08-16 Fri 08:39]--[2024-08-16 Fri 17:39] =>  9:00
:END:
****** Test:
1. navigate through content without overlay states being affected
****** 0. Redefine Router, let everything point to DM
****** 1. Add content liveview
- all 3 added
- [LOW IMPORTANCE]

  TOVERIFY: seems like it's correct to preload the stream content @ the liveview then pass it onto a live_component? ([[https://elixirforum.com/t/how-to-use-phoenix-1-7-streams-on-livecomponent-preload/54454][ref]])
  my own implementations:

  PRELOAD @
  #+begin_src elixir
# preload @ display_manager/display_live
  defp apply_action(
         %Socket{} = socket,
         :show_chapters,
         %{"source_title" => source_title} =
           params
       ) do

    [%Chapter{source: src} | _] = chapters = Written.get_chapters_by_src(source_title)
    socket
    |> assign(:content_action, :show_chapters)
    |> assign(:page_title, to_title_case(src.title))
    |> assign(:source, src)
    |> assign(:meta, %{
      title: to_title_case(src.title),
      description: "Explore the #{to_title_case(src.title)}",
      type: "website",
      image: url(~p"/og/#{VyasaWeb.OgImageController.get_by_binding(%{source: src})}"),
      url: url(socket, ~p"/explore/#{src.title}")
    })
    |> stream_configure(:chapters, dom_id: &"Chapter-#{&1.no}")
    |> stream(:chapters, chapters |> Enum.sort_by(fn chap -> chap.no end))
  end
  #+end_src

  INJECT INTO LIVE_COMPONENT:

  #+begin_src elixir
<%= if @content_action == :show_chapters do%>
    <div>
        SHOW CHAPTERS
   <.live_component
      module={VyasaWeb.Content.Chapters}
      id={"content-sources"}
      source={@source}
      chapters={@streams.chapters}
      />
    </div>
<% end %>
  #+end_src

****** 2. defdelegate or add contentFetcher to separate out the repo logic
SKIPPED this, not too important
****** 3. test the push_patch() for this
WORKSSSS

```
